File: Ontology/ObjectHasSelf.cs
namespace Ontology
{
    public class ObjectHasSelf:
        ObjectPropertyRestriction,
        IObjectHasSelf
    {
        public ObjectHasSelf(
            IObjectPropertyExpression objectPropertyExpression
            ) : base(objectPropertyExpression)
        {
        }

        public override void Accept(
            IClassExpressionVisitor visitor
            ) => visitor.Visit(this);

        public override bool Evaluate(
            IClassMembershipEvaluator evaluator,
            object individual
            ) => evaluator.Evaluate(
                this,
                individual);
    }
}

File: Web/ClientApp/src/app/Ontology/ObjectHasSelf.ts
import { IClassExpressionSelector } from './IClassExpressionSelector';
import { IClassExpressionVisitor } from "./IClassExpressionVisitor";
import { IObjectHasSelf } from "./IObjectHasSelf";
import { IObjectPropertyExpression } from "./IPropertyExpression";
import { ObjectPropertyRestriction } from "./ObjectPropertyRestriction";

export class ObjectHasSelf
    extends ObjectPropertyRestriction
    implements IObjectHasSelf
{
    public constructor(
        objectPropertyExpression: IObjectPropertyExpression
        )
    {
        super(objectPropertyExpression)
    }

    Accept(
        visitor: IClassExpressionVisitor
        )
    {
        visitor.ObjectHasSelf(this);
    }

    Select<TResult>(
        selector: IClassExpressionSelector<TResult>
        ): TResult
    {
        return selector.ObjectHasSelf(this);
    }
}
